processing/src/main/java/io/druid/query/select/SelectQueryQueryToolChest.java

```diff
@@ -359,9 +360,19 @@ public Object apply(final Result<SelectResultValue> input)
     }
 
     final Granularity granularity = query.getGranularity();
+    final Iterable<String> filteredPagingKeys = Iterables.filter(
+        paging.keySet(), new Predicate<String>()
+        {
+          @Override
+          public boolean apply(String input)
+          {
+            return input.startsWith(String.format("%s_", dataSource));
```

Comment:
DataSegmentUtils also doesn't seem to be perfect, this test case won't pass (it matches datasources sometimes when it shouldn't). I'm not sure how easy it is to fix that, but it'd have to be fixed to make this as robust as possible to datasource names.

```
  @Test
  public void testDataSourceWithUnderscore3()
  {
    final String datasource = "datasource";
    final String datasource2000 = "datasource_2000-01-01T00:00:00.000Z";

    final SegmentIdentifierParts descNoTrail = new SegmentIdentifierParts(
        datasource,
        new Interval("2015-01-02/2015-01-03"),
        "ver",
        null
    );

    final SegmentIdentifierParts descWithTrail = new SegmentIdentifierParts(
        datasource,
        new Interval("2015-01-02/2015-01-03"),
        "ver",
        "0"
    );

    final SegmentIdentifierParts desc2000NoTrail = new SegmentIdentifierParts(
        datasource2000,
        new Interval("2015-01-02/2015-01-03"),
        "ver",
        null
    );

    final SegmentIdentifierParts desc2000WithTrail = new SegmentIdentifierParts(
        datasource2000,
        new Interval("2015-01-02/2015-01-03"),
        "ver",
        "0"
    );

    Assert.assertEquals(
        "datasource_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver",
        descNoTrail.toString()
    );
    Assert.assertEquals(
        "datasource_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver_0",
        descWithTrail.toString()
    );
    Assert.assertEquals(
        datasource2000 + "_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver",
        desc2000NoTrail.toString()
    );
    Assert.assertEquals(
        datasource2000 + "_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver_0",
        desc2000WithTrail.toString()
    );

    Assert.assertEquals(
        descNoTrail,
        DataSegmentUtils.valueOf(datasource, descNoTrail.toString())
    );
    Assert.assertEquals(
        descWithTrail,
        DataSegmentUtils.valueOf(datasource, descWithTrail.toString())
    );
    Assert.assertNull(DataSegmentUtils.valueOf(datasource, desc2000NoTrail.toString()));
    Assert.assertNull(DataSegmentUtils.valueOf(datasource, desc2000WithTrail.toString()));

    Assert.assertEquals(
        desc2000NoTrail,
        DataSegmentUtils.valueOf(datasource2000, desc2000NoTrail.toString())
    );
    Assert.assertEquals(
        desc2000WithTrail,
        DataSegmentUtils.valueOf(datasource2000, desc2000WithTrail.toString())
    );
    Assert.assertNull(DataSegmentUtils.valueOf(datasource2000, descNoTrail.toString()));
    Assert.assertNull(DataSegmentUtils.valueOf(datasource2000, descWithTrail.toString()));
  }
```